### PR TITLE
Support Chaining for some of functionalities of `nn.Module` (#885)

### PR DESCRIPTION
**Pull request by amirhossein-razlighi**
## Proposed changes

Closes issue #885 . This PR adds chaining support for some of functionalities in `nn.Module`. In this PR, functions like `.freeze()`, `.update()`, `train()`, etc. return the instance of that Module, instead of returning None. So, operations like these are possible:
```
import mlx.core as mx
import mlx.nn as nn

m = nn.Sequential(nn.Linear(2, 2), nn.ReLU(), nn.Linear(2, 1))
m.freeze().unfreeze().train()
```

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/layers/base.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def load_weights(
         self,
         file_or_weights: Union[str, List[Tuple[str, mx.array]]],
         strict: bool = True,
-    ):
```
Comment (awni) [inline python/mlx/nn/layers/base.py:149]: Why quote Module here? Can you not just do `-> Module` directly?

File: python/mlx/nn/layers/base.py [hunk 2]
```diff
@@ -540,6 +564,9 @@ def unfreeze(
                 ``module.unfreeze(keys="bias")``.
             strict (bool, optional): If set to ``True`` validate that the passed keys exist.
                 Default: ``False``.
+
+        Returns:
+            The module instance with unfreezed parameters.
```
Comment (awni) [inline python/mlx/nn/layers/base.py:569]:
```suggestion
            The module instance with unfrozen parameters.
```

File: python/mlx/nn/layers/base.py [hunk 3]
```diff
@@ -488,6 +508,9 @@ def freeze(
                 ``module.freeze(keys="bias")``.
             strict (bool, optional): If set to ``True`` validate that the passed keys exist.
                 Default: ``False``.
+
+        Returns:
+            The module instance with freezed parameters.
```
Comment (awni) [inline python/mlx/nn/layers/base.py:513]:
```suggestion
            The module instance with frozen parameters.
```

**Review by awni — COMMENTED** [review]
Do you think this change introduces some ambiguity which might cause confusion?

Like:

`m1 = m.freeze()`

Does one expect that `m1` refers to the same module as `m`?

File: python/mlx/nn/layers/base.py (hunk 1 above)
Comment (amirhossein-razlighi) [reply python/mlx/nn/layers/base.py:149]: No. If I do `-> Module` or `-> self`, since the class is not defined completely yet, it gives error `Module is not defined`. So, we should use python's "forward referencing" that is mentioning the class name inside quotes

**Comment by amirhossein-razlighi** [comment]
I do not think this change, causes ambiguity. Since one may expect that every valid operation on an instance of `nn.Module` has the ability to support chaining. And some operations like `.freeze()` which return `None` before, right now return the module instance itself, so no conflicts are caused. And yes, I think in case of `m1 = m.freeze()`, one may expect `m1` refers to `m` module, but with frozen parameters.

**Comment by amirhossein-razlighi** [comment]
> Do you think this change introduces some ambiguity which might cause confusion?
> 
> Like:
> 
> `m1 = m.freeze()`
> 
> Does one expect that `m1` refers to the same module as `m`?

I do not think this change, causes ambiguity. Since one may expect that every valid operation on an instance of `nn.Module` has the ability to support chaining. And some operations like `.freeze()` which return `None` before, right now return the module instance itself, so no conflicts are caused. And yes, I think in case of `m1 = m.freeze()`, one may expect `m1` refers to `m` module, but with frozen parameters.

File: python/mlx/nn/layers/base.py (hunk 1 above)
Comment (awni) [reply python/mlx/nn/layers/base.py:149]: I see, that makes sense.

**Comment by amirhossein-razlighi** [comment]
> > Do you think this change introduces some ambiguity which might cause confusion?
> > Like:
> > `m1 = m.freeze()`
> > Does one expect that `m1` refers to the same module as `m`?
> 
> I do not think this change, causes ambiguity. Since one may expect that every valid operation on an instance of `nn.Module` has the ability to support chaining. And some operations like `.freeze()` which return `None` before, right now return the module instance itself, so no conflicts are caused. And yes, I think in case of `m1 = m.freeze()`, one may expect `m1` refers to `m` module, but with frozen parameters.

Do you think it's okay? In this case, I think this PR is almost finished and ready to go 🚀

**Comment by amirhossein-razlighi** [comment]
> > Do you think this change introduces some ambiguity which might cause confusion?
> > Like:
> > `m1 = m.freeze()`
> > Does one expect that `m1` refers to the same module as `m`?
> 
> I do not think this change, causes ambiguity. Since one may expect that every valid operation on an instance of `nn.Module` has the ability to support chaining. And some operations like `.freeze()` which return `None` before, right now return the module instance itself, so no conflicts are caused. And yes, I think in case of `m1 = m.freeze()`, one may expect `m1` refers to `m` module, but with frozen parameters.

@awni 
Do you think it's okay? In this case, I think this PR is almost finished and ready to go 🚀

File: python/mlx/nn/layers/base.py [hunk 4]
```diff
@@ -159,6 +159,9 @@ def load_weights(
               only the weights actually contained in the model are loaded and
               shapes are not checked. Default: ``True``.
 
+        Returns:
+            The model with the updated weights.
```
Comment (angeloskath) [inline python/mlx/nn/layers/base.py:163]: To further reduce the risk of confusion perhaps we can reword this to "The module instance after updating the weights", just to reinforce that it is the same module being returned.

**Review by angeloskath — APPROVED** [review]
I think this looks good. I would reword the return docs but in general I don't think there will be much confusion.

**Comment by awni** [comment]
I like @angeloskath suggestion, would you mind making the change @amirhossein-razlighi ? Then we can run the tests and merge.

File: python/mlx/nn/layers/base.py (hunk 4 above)
Comment (amirhossein-razlighi) [reply python/mlx/nn/layers/base.py]: Sure. Thanks for the comment!

**Comment by amirhossein-razlighi** [comment]
> I like @angeloskath suggestion, would you mind making the change @amirhossein-razlighi ? Then we can run the tests and merge.

Done ✅

**Comment by angeloskath** [comment]
Sorry to be a stickler but can you also change the rest of the docs. ie "The module instance after updating the training mode".

**Comment by amirhossein-razlighi** [comment]
> Sorry to be a stickler but can you also change the rest of the docs. ie "The module instance after updating the training mode".

Done ✅

**Comment by awni** [comment]
Thanks for adding this!
